diff --git git/arch/arm/mach-omap2/board-overo.c-orig git/arch/arm/mach-omap2/board-overo.c
index 95efc77..6d6ed97 100644
--- git/arch/arm/mach-omap2/board-overo.c-orig
+++ git/arch/arm/mach-omap2/board-overo.c
@@ -225,6 +225,11 @@ static inline void __init overo_init_smsc911x(void) { return; }
 static int lcd_enabled;
 static int dvi_enabled;
 
+#if defined(CONFIG_PANEL_LGPHILIPS_LB035Q02) || \
+	defined(CONFIG_PANEL_LGPHILIPS_LB035Q02_MODULE) || \
+	defined(CONFIG_PANEL_SAMSUNG_LTE430WQ_F0C) || \
+	defined(CONFIG_PANEL_SAMSUNG_LTE430WQ_F0C_MODULE)
+
 #define OVERO_GPIO_LCD_EN 144
 #define OVERO_GPIO_LCD_BL 145
 
@@ -244,6 +249,9 @@ static void __init overo_display_init(void)
 		printk(KERN_ERR "could not obtain gpio for "
 					"OVERO_GPIO_LCD_BL\n");
 }
+#else
+static void __init overo_display_init(void) { return; }
+#endif
 
 static int overo_panel_enable_dvi(struct omap_dss_device *dssdev)
 {
@@ -277,6 +285,10 @@ static struct omap_dss_device overo_tv_device = {
 	.phy.venc.type = OMAP_DSS_VENC_TYPE_SVIDEO,
 };
 
+#if defined(CONFIG_PANEL_LGPHILIPS_LB035Q02) || \
+	defined(CONFIG_PANEL_LGPHILIPS_LB035Q02_MODULE) || \
+	defined(CONFIG_PANEL_SAMSUNG_LTE430WQ_F0C) || \
+	defined(CONFIG_PANEL_SAMSUNG_LTE430WQ_F0C_MODULE)
 static int overo_panel_enable_lcd(struct omap_dss_device *dssdev)
 {
 	if (dvi_enabled) {
@@ -296,6 +308,7 @@ static void overo_panel_disable_lcd(struct omap_dss_device *dssdev)
 	gpio_set_value(OVERO_GPIO_LCD_BL, 0);
 	lcd_enabled = 0;
 }
+#endif
 
 #if defined(CONFIG_PANEL_LGPHILIPS_LB035Q02) || \
 	defined(CONFIG_PANEL_LGPHILIPS_LB035Q02_MODULE)
