diff --git git/arch/arm/plat-omap/mcbsp.c-orig git/arch/arm/plat-omap/mcbsp.c
index 1c2a147..340d647 100644
--- git/arch/arm/plat-omap/mcbsp.c-orig
+++ git/arch/arm/plat-omap/mcbsp.c
@@ -210,6 +210,12 @@ void omap_mcbsp_config(unsigned int id, const struct omap_mcbsp_reg_cfg *config)
 	dev_dbg(mcbsp->dev, "Configuring McBSP%d  phys_base: 0x%08lx\n",
 			mcbsp->id, mcbsp->phys_base);
 
+	dev_dbg(mcbsp->dev, "    io_type: %d\n", mcbsp->io_type);
+	dev_dbg(mcbsp->dev, "    tx_irq: %d\n", mcbsp->tx_irq);
+	dev_dbg(mcbsp->dev, "    rx_irq: %d\n", mcbsp->rx_irq);
+	dev_dbg(mcbsp->dev, "    dma_rx_sync: %d\n", mcbsp->dma_rx_sync);
+	dev_dbg(mcbsp->dev, "    dma_tx_sync: %d\n", mcbsp->dma_tx_sync);
+
 	/* We write the given config */
 	MCBSP_WRITE(mcbsp, SPCR2, config->spcr2);
 	MCBSP_WRITE(mcbsp, SPCR1, config->spcr1);
