diff --git perl-5.10.1/Configure-orig perl-5.10.1/Configure
index 01fa3c0..637ba32 100755
--- perl-5.10.1/Configure-orig
+++ perl-5.10.1/Configure
@@ -1307,11 +1307,13 @@ loclibpth="/usr/local/lib /opt/local/lib /usr/gnu/lib"
 loclibpth="$loclibpth /opt/gnu/lib /usr/GNU/lib /opt/GNU/lib"
 
 : general looking path for locating libraries
-glibpth="/lib /usr/lib $xlibpth"
+glibpth="/lib /usr/lib /lib64 /usr/lib64 $xlibpth"
 glibpth="$glibpth /usr/ccs/lib /usr/ucblib /usr/local/lib"
 test -f /usr/shlib/libc.so && glibpth="/usr/shlib $glibpth"
 test -f /shlib/libc.so     && glibpth="/shlib $glibpth"
 test -d /usr/lib64         && glibpth="$glibpth /lib64 /usr/lib64 /usr/local/lib64"
+test -f /usr/lib/i386-linux-gnu/libc.so && glibpth="/usr/lib/i386-linux-gnu $glibpth"
+test -f /usr/lib/x86_64-linux-gnu/libc.so && glibpth="/usr/lib/x86_64-linux-gnu $glibpth"
 
 : Private path used by Configure to find libraries.  Its value
 : is prepended to libpth. This variable takes care of special
