diff --git git/arch/arm/mach-omap2/board-overo.c-orig git/arch/arm/mach-omap2/board-overo.c
index 9ee600c..ebd5185 100644
--- git/arch/arm/mach-omap2/board-overo.c-orig
+++ git/arch/arm/mach-omap2/board-overo.c
@@ -633,12 +633,29 @@ static struct i2c_board_info __initdata overo_i2c_boardinfo[] = {
 	},
 };
 
+#define LM3553_SLAVE_ADDRESS	0x53
+#define OV3640_I2C_ADDR		(0x78 >> 1)
+int ov3640_platform_data;
+int lm3553_platform_data;
+
+static struct i2c_board_info __initdata econ_cam_boardinfo[] = {
+       {
+               I2C_BOARD_INFO("ov3640", OV3640_I2C_ADDR),
+               .platform_data = &ov3640_platform_data,
+       },
+       {
+               I2C_BOARD_INFO("lm3553", LM3553_SLAVE_ADDRESS),
+               .platform_data = &lm3553_platform_data,
+       },
+};
+
 static int __init overo_i2c_init(void)
 {
 	omap_register_i2c_bus(1, 2600, overo_i2c_boardinfo,
 			ARRAY_SIZE(overo_i2c_boardinfo));
 	/* i2c2 pins are used for gpio */
-	omap_register_i2c_bus(3, 400, NULL, 0);
+	omap_register_i2c_bus(3, 100, econ_cam_boardinfo, 
+			ARRAY_SIZE(econ_cam_boardinfo));
 	return 0;
 }
 
