diff --git git/drivers/media/video/Kconfig-orig git/drivers/media/video/Kconfig
index f6e4d04..63a62ef 100644
--- git/drivers/media/video/Kconfig-orig
+++ git/drivers/media/video/Kconfig
@@ -774,6 +774,13 @@ config VIDEO_CAFE_CCIC
 	  CMOS camera controller.  This is the controller found on first-
 	  generation OLPC systems.
 
+config VIDEO_OMAP3
+       bool "OMAP2/OMAP3 V4L2-DSS drivers"
+       depends on VIDEO_DEV && (ARCH_OMAP24XX || ARCH_OMAP34XX || ARCH_OMAP3430)
+       default y
+       ---help---
+        V4L2 DSS driver support for OMAP2/3 based boards.
+
 config SOC_CAMERA
 	tristate "SoC camera support"
 	depends on VIDEO_V4L2 && HAS_DMA && I2C
