diff --git a/git/include/asm-arm/arch-omap3/mem.h.orig b/git/include/asm-arm/arch-omap3/mem.h
index 5b9ac75..39b48d4 100644
--- git/include/asm-arm/arch-omap3/mem.h.orig
+++ git/include/asm-arm/arch-omap3/mem.h
@@ -196,11 +196,11 @@ enum {
 #define ONENAND_GPMC_CONFIG6	0x1F060000

 #define NET_GPMC_CONFIG1	0x00001000
-#define NET_GPMC_CONFIG2	0x001e1e01
-#define NET_GPMC_CONFIG3	0x00080300
-#define NET_GPMC_CONFIG4	0x1c091c09
-#define NET_GPMC_CONFIG5	0x04181f1f
-#define NET_GPMC_CONFIG6	0x00000FCF
+#define NET_GPMC_CONFIG2	0x00080701
+#define NET_GPMC_CONFIG3	0x00020201
+#define NET_GPMC_CONFIG4	0x08030703
+#define NET_GPMC_CONFIG5	0x00060908
+#define NET_GPMC_CONFIG6	0x89030000
 #define NET_GPMC_CONFIG7	0x00000f6c

 /* max number of GPMC Chip Selects */


