diff --git git/drivers/mfd/twl-core.c-orig git/drivers/mfd/twl-core.c
index 07af8e0..55df70d 100644
--- git/drivers/mfd/twl-core.c-orig
+++ git/drivers/mfd/twl-core.c
@@ -208,6 +208,7 @@
 
 /* Few power values */
 #define R_CFG_BOOT			0x05
+#define R_GPBR1				0x0C
 #define R_PROTECT_KEY			0x0E
 
 /* access control values for R_PROTECT_KEY */
@@ -215,6 +216,10 @@
 #define KEY_UNLOCK2			0xec
 #define KEY_LOCK			0x00
 
+/* MADC clock values for R_GPBR1 */
+#define MADC_HFCLK_EN			0x80
+#define DEFAULT_MADC_CLK_EN		0x10
+
 /* some fields in R_CFG_BOOT */
 #define HFCLK_FREQ_19p2_MHZ		(1 << 0)
 #define HFCLK_FREQ_26_MHZ		(2 << 0)
@@ -932,6 +937,10 @@ static void clocks_init(struct device *dev,
 
 	e |= unprotect_pm_master();
 	/* effect->MADC+USB ck en */
+	if (twl_has_madc())
+		e |= twl_i2c_write_u8(TWL4030_MODULE_INTBR,
+			MADC_HFCLK_EN | DEFAULT_MADC_CLK_EN, R_GPBR1);
+
 	e |= twl_i2c_write_u8(TWL_MODULE_PM_MASTER, ctrl, R_CFG_BOOT);
 	e |= protect_pm_master();
 
